Corgibytes.Freshli.Cli.Test/CommandOptions/ScanCommandOptionsTest.cs

```diff
@@ -0,0 +1,83 @@
+
+            outputStrategyTypes.Should().NotBeEmpty()
+            .And.Equal(outputStrategyTypesFromAlias);
+
+            path.Should().NotBeNull();
+            path.FullName.Should().NotBeEmpty()
+            .And.Be(expectedPath);
+
+            formatType.Should()
+            .Be(expectedFormat);
+
+            outputStrategyTypes.Should().NotContainNulls()
+            .And.NotContain(OutputStrategyType.None);
+
+            //Veryfy all required ones are present
+            expectedOutput.Should().BeSubsetOf(outputStrategyTypes);
+
+            //Veryfy there aren't extra output stragetegies added
+            outputStrategyTypes.Should().BeSubsetOf(expectedOutput);
+        }
+
+        public static IEnumerable<object[]> ScanOptionsArgs =>
+              new List<object[]>
+              {
+                    new object[] { new string[] { "scan", TempPath, "--format", "json", "--output", "console"}, TempPath, FormatType.Json, new List<OutputStrategyType>() { OutputStrategyType.Console } },
+                    new object[] { new string[] { "scan", TempPath, "--Format", "JSON", "--output", "CONSOLE" }, TempPath, FormatType.Json, new List<OutputStrategyType>() { OutputStrategyType.Console } },
+                    new object[] { new string[] { "scan", TempPath, "--format", "csv", "--output", "file", "--output", "console" }, TempPath, FormatType.Csv, new List<OutputStrategyType>() { OutputStrategyType.File, OutputStrategyType.Console }},
+                    new object[] { new string[] { "scan", TempPath, "--format", "yaml", "--output", "file" }, TempPath, FormatType.Yaml, new List<OutputStrategyType>() { OutputStrategyType.File }},
+                    new object[] { new string[] { "scan", TempPath, "-f", "json", "-o", "console" }, TempPath, FormatType.Json, new List<OutputStrategyType>() { OutputStrategyType.Console } },
+                    new object[] { new string[] { "scan", TempPath, "-f", "csv", "-o", "file", "-o", "console" }, TempPath, FormatType.Csv, new List<OutputStrategyType>() { OutputStrategyType.File, OutputStrategyType.Console }},
+                    new object[] { new string[] { "scan", TempPath, "-f", "Csv", "-o", "FILE", "-o", "console" }, TempPath, FormatType.Csv, new List<OutputStrategyType>() { OutputStrategyType.File, OutputStrategyType.Console }},
+                    new object[] { new string[] { "scan", TempPath, "-f", "yaml", "-o", "console" }, TempPath, FormatType.Yaml, new List<OutputStrategyType>() { OutputStrategyType.Console } },
+                    new object[] { new string[] { "scan", TempPath, "--format", "yaml", "-o", "file" }, TempPath, FormatType.Yaml, new List<OutputStrategyType>() { OutputStrategyType.File }},
+                    new object[] { new string[] { "scan", TempPath, "-f", "yaml", "--output", "console","-o", "file" }, TempPath, FormatType.Yaml, new List<OutputStrategyType>() { OutputStrategyType.Console, OutputStrategyType.File } },
```

Comment:
> * Do we need all the interfaces?  It appears that some of the interfaces are only used in one or two spots.  Also sometimes there is a base class and a interface.  Are both needed?  My guess is the only reason they are created is to support dependency injection framework.
> *

My responses below:

1)

> Some formatting requests in the inline comments. My bigger questions are:
> 
> 1. Do we need all the interfaces?  It appears that some of the interfaces are only used in one or two spots.  Also sometimes there is a base class and a interface.  Are both needed?  My guess is the only reason they are created is to support dependency injection framework.
> 2. Do we need to register every class when configuring the services or just one or two?
> 3. I don't fully understand the [Host](https://docs.microsoft.com/en-us/dotnet/core/extensions/generic-host) in the `Program.cs` file.  I think it just helps with setup but can you please clarify.  I assume it is not creating a Windows Service or running as a server.
> 4. I don't think the `IoCCommandRunnerFactory.cs` is actually a factory.  At least not factories I remember but please correct me if I'm wrong.
> 5. All user output from the CLI, such as help, should be multilingual.
> 6. I don't think OutputStrategies needs the "Strategies/Strategy" part.  For example "ConsoleOutputStrategy" should just be "ConsoleOutput".
> 7. Can we remove the Auth classes until they are implemented?  In the inline comments I asked about the help command but I just noticed that is solved by having the `--help` option.
> 8. Please update the ReadMe with the new command format.
> 9. Please add some documentation about the high level structure of the CLI.  For example, discuss the CommandOptions, Commands, etc.
> 
> Basically I don't understand the new way to write console apps in .NET. I also want to check that the extra complexity of the extra interfaces and IoC is worth it. We might have to pair for you to explain to an old programmer all this new way of writing console applications.
> 
> Finally when I try to run the application I get the error below. Also a bunch of log information is printed out which I don't think should be there.
> 
> ```
> > Corgibytes.Freshli.Cli.exe scan https://github.com/corgibytes/freshli-fixture-ruby-nokotest
> info: Microsoft.Hosting.Lifetime[0]
>       Application started. Press Ctrl+C to shut down.
> info: Microsoft.Hosting.Lifetime[0]
>       Hosting environment: Production
> info: Microsoft.Hosting.Lifetime[0]
>       Content root path: C:\Repos\freshli-cli\Corgibytes.Freshli.Cli\bin\Debug\net5.0
> [Command Execution Invocation Started  - ParseResult: [ Corgibytes.Freshli.Cli [ scan <https://github.com/corgibytes/freshli-fixture-ruby-nokotest> *[ --format <Json> ] *[ --output <Console> ] ] ] ]
> CliOutput.ScanCommand_ScanCommand_Executing_scan_command_handler
> [Unhandled Exception - ParseResult: [ Corgibytes.Freshli.Cli [ scan <https://github.com/corgibytes/freshli-fixture-ruby-nokotest> *[ --format <Json> ] *[ --output <Console> ] ] ]] - Exception has been thrown by the target of an invocation. - Take a look at the log for detailed information.
> info: Microsoft.Hosting.Lifetime[0]
>       Application is shutting down...
> ```
> 
> If I run the CLI with not arguments I get the help message but still get the extra log information:
> 
> ```
> > Corgibytes.Freshli.Cli.exe
> info: Microsoft.Hosting.Lifetime[0]
>       Application started. Press Ctrl+C to shut down.
> info: Microsoft.Hosting.Lifetime[0]
>       Hosting environment: Production
> info: Microsoft.Hosting.Lifetime[0]
>       Content root path: C:\Repos\freshli-cli\Corgibytes.Freshli.Cli\bin\Debug\net5.0
> [Command Execution Invocation Started  - ParseResult: ![ Corgibytes.Freshli.Cli ] ]
> [Command Execution Invocation Ended - ParseResult: ![ Corgibytes.Freshli.Cli ] ]
> info: Microsoft.Hosting.Lifetime[0]
>       Application is shutting down...
> Required command was not provided.
> 
> Corgibytes.Freshli.Cli
>   Root Command
> 
> Usage:
>   Corgibytes.Freshli.Cli [options] [command]
> 
> Options:
>   -?, -h, --help  Show help and usage information
>   --version       Show version information
> 
> Commands:
>   scan <path>  Scan command returns metrics results for given local repository path
>   auth         Auth Command
> ```
> 
> Let me know if you get the same error. If not it might be an issue on my machine/setup.

My responses below:

1) Yes, we need interfaces because that is the way to decouple the contract from the implementation. This way we do not force class relationships and we make the code fully independent and testable.  The case you mentioned about the abstract class and the interface, yes we need both, the interface for the same previous reason and the abstract class because I have some common code shared among children that I do not want to duplicate. This is not related to IoC.

2) I do not understand this question.

3) It is not creating any service. It is the way of writing console applications in dotnetcore. It allows you to configure multiple things on startup.  You can see an example here https://github.com/dotnet/command-line-api. This is an example in the library that we use for parsing.

4) I am implementing the Abstract Factory pattern here (https://www.dofactory.com/net/abstract-factory-design-pattern). I would like to avoid using case or multiple ifs statements. It could probably lead to methods with lots of lines of codes. 

5) I am not sure if the help command language can be configured. I am going to do some research.

6) I am using the word strategy because I am implementing the Strategy Design Pattern. I think that makes people easy to understand what is happening. You can see the pattern here . https://www.dofactory.com/net/strategy-design-pattern

7) I would keep the auth classes to show how to add a new command. I already made it to throw a NotImplementedException if called.

8) 9) Yes, will update readme. Regarding documentation I had already created the activity diagram that I uploaded to notion. I can make it available in the repository and add more documentation.

I will allow the command line to use the repository public URL instead of local path (or keep both). For some reason, there was a misunderstood in that point. I was sure that we said we were going to allow local paths only but I also can be wrong, of course hehe.

Regarding IoC, I do not see an extra complexity in there to be honest. From my point of view, this is the way we should go to make the code fully decoupled and the way I do programming. We need to keep it decoupled as much as possible. We should avoid instantiating classes ourselves.